src/Sarif/FileEncodings.cs

```diff
@@ -0,0 +1,137 @@
+    files (the “Software”), to deal in the Software without
+    restriction, including without limitation the rights to use,
+    copy, modify, merge, publish, distribute, sublicense, and/or sell
+    copies of the Software, and to permit persons to whom the
+    Software is furnished to do so, subject to the following
+    conditions:
+
+    The above copyright notice and this permission notice shall be
+    included in all copies or substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND,
+    EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+    OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+    NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+    HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+    WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+    OTHER DEALINGS IN THE SOFTWARE. 
+
+    License: http://www.opensource.org/licenses/mit-license.php
+    Website: https://github.com/khalidsalomao/SimpleHelpers.Net
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+
+namespace Microsoft.CodeAnalysis.Sarif
+{
+    public class FileEncoding
+    {
+        /// <summary>
+        /// Detects if contains textual data.
+        /// </summary>
+        /// <param name="rawData">The raw data.</param>
+        public static bool CheckForTextualData(byte[] rawData, out Encoding encoding)
+        {
+            return CheckForTextualData(rawData, 0, rawData.Length, out encoding);
+        }
+
+        /// <summary>
+        /// Detects if contains textual data.
+        /// </summary>
+        /// <param name="rawData">The raw data.</param>
+        /// <param name="start">The start.</param>
+        /// <param name="count">The count.</param>
+        public static bool CheckForTextualData(byte[] rawData, int start, int count, out Encoding encoding)
+        {
+            encoding = null;
+
+            if (rawData.Length < count || count < 4 || start + 1 >= count)
+            {
+                return true;
+            }
+
+            if (CheckForByteOrderMark(rawData, out encoding, start))
+            {
+                return true;
+            }
+
+            // http://stackoverflow.com/questions/910873/how-can-i-determine-if-a-file-is-binary-or-text-in-c
+            // http://www.gnu.org/software/diffutils/manual/html_node/Binary.html
+            // count the number od null bytes sequences
+            // considering only sequences of 2 0s: "\0\0" or control characters below 10
+            int nullSequences = 0;
+            int controlSequences = 0;
+            for (int i = start + 1; i < count; i++)
+            {
+                if (rawData[i - 1] == 0 && rawData[i] == 0)
+                {
+                    if (++nullSequences > 1)
```

Comment:
>>  [](http://example.com/codeflow?start=40&length=2)

`++nullSequences > 1` requires finding of 2 '\0\0' sequences. 
Only 1 should be enough? `>= 1` or `> 0`

Reply:
The problem, I believe, is that some BOM sequences has a '\0' character (but no more than 1). So your heuristic might make use believe that a text file with a BOM is binary.